### PR TITLE
Add row locking grammer.

### PR DESCRIPTION
**Pull request by chrisxu333**
I added support for a full functional row locking grammar including:

- FOR (NO KEY) UPDATE (OF), FOR (KEY) SHARE (OF)
- NOWAIT and SKIP LOCKED specifier

### CODE REVIEW

**Comment by Bouncner** [comment]
Wasn't aware of this feature. Just for reference: this is the postgres introducing this functionality https://github.com/postgres/postgres/commit/0ac5ad5134f2769ccbaefec73844f8504c4d6182

**Comment by Bouncner** [comment]
@chrisxu333 Are you aware of any other database that uses a similar functionality and syntax? Or is this postgres-specific?

**Comment by chrisxu333** [comment]
@Bouncner Speaking of SELECT FOR UPDATE (OF) and SELECT FOR SHARE (OF), I believe most well-known databases support this, like MySQL, Oracle, CockroachDB, Postgres and many more, and I believe they all support NOWAIT and SKIP LOCKED flag. And for the NO KEY and KEY option, I know for sure that CockroachDB and Postgres support this but not sure about others.

**Comment by mweisgut** [comment]
Related Postgres references:
- https://www.postgresql.org/docs/14/explicit-locking.html
- https://www.postgresql.org/docs/14/sql-select.html

File: src/parser/bison_parser.y [hunk 1]
```diff
@@ -1114,6 +1120,49 @@ alias : AS IDENTIFIER { $$ = new Alias($2); }
...
+ }
+ |/* empty */{ $$ = nullptr; };
+
+ for_locking_clause_waiting_policy: SKIP LOCKED{
+   $$ = 0;
+ }
+ | NOWAIT{
+   $$ = 1;
+ }
+ |/* empty */{ $$ = 2; };
```
Comment (mweisgut) [inline src/parser/bison_parser.y:1164]: Unfortunately, integers as representatives for certain modes are not particularly descriptive. As mentioned above, I think using enumeration values would increase the readability.

File: src/sql/SelectStatement.h [hunk 2]
```diff
@@ -96,6 +105,7 @@ struct SelectStatement : SQLStatement {
   std::vector<OrderDescription*>* order;
   std::vector<WithDescription*>* withDescriptions;
   LimitDescription* limit;
+  ForLockingClause* lockings;
```
Comment (mweisgut) [inline src/sql/SelectStatement.h:108]: The implementation of this pull request only supports one locking clause. However, multiple can be passed. Thus, we should store multiple locking clauses here.

https://www.postgresql.org/docs/14/sql-select.html

File: src/parser/bison_parser.y [hunk 3]
```diff
@@ -746,25 +749,28 @@ select_within_set_operation_no_parentheses : select_clause { $$ = $1; }
...
   // Limit could have been set by TOP.
-  if ($3 != nullptr) {
+  if ($4 != nullptr) {
     delete $$->limit;
-    $$->limit = $3;
+    $$->limit = $4;
   }
 }
-| select_clause set_operator select_within_set_operation opt_order opt_limit {
+| select_clause set_operator select_within_set_operation opt_order for_locking_clause opt_limit {
```
[the remark this answers is not in the document]
Comment (mweisgut) [reply src/parser/bison_parser.y:764]: Furthermore, multiple locking clauses are possible in a select statement.

**Review by mweisgut — CHANGES_REQUESTED** [review]
Thank you, @chrisxu333, for this patch. We want to add your patch to the parser. To do so, we kindly ask you to address the comments listed below.

Besides these aspects, tests are entirely missing. Could you please add tests to `select_tests.cpp`, add correct select statements with the added keywords and right syntax to `queries-good.sql`, and statements with the added keywords buy incorrect syntax to `queries-bad.sql`?
Furthermore, could you please run the `fromat.sh` script from the project's root directory to match the code base's formatting?

File: src/sql/SelectStatement.h (hunk 2 above)
Comment (chrisxu333) [reply src/sql/SelectStatement.h:108]: Thank you for accepting my patch into the project. I'm working on refining it now. One question here: what do you mean by saying multiple locking clause can be passed here, could you give an SQL query example of this?

Comment (chrisxu333) [reply src/sql/SelectStatement.h:108]: I think I figured out what you meant. I'm working on it right now.

Comment (mweisgut) [reply src/sql/SelectStatement.h:108]: The following query is an example with multiple locking clauses:
```
CREATE TABLE t1 (c1 INT);
CREATE TABLE t2 (c1 INT);
INSERT INTO t1 (c1) VALUES (5);
INSERT INTO t2 (c1) VALUES (3);
SELECT t1.c1 AS t1c1, t2.c1 AS t2c1 FROM t1, t2
FOR UPDATE OF t1 NOWAIT
FOR NO KEY UPDATE OF t2 NOWAIT;
```

Verified with PostgreSQL on https://sqliteonline.com/.

**Comment by chrisxu333** [comment]
@mweisgut I‘ve addressed the issues you brought up in your comment in the latest commit, including testcases, format issue, naming conventions as well as multiple locking clause support and such. Please take a look and let me know if anything else needs to be changed. Thank you!

File: src/sql/SelectStatement.h [hunk 4]
```diff
@@ -10,6 +10,9 @@ enum OrderType { kOrderAsc, kOrderDesc };
 
 enum SetType { kSetUnion, kSetIntersect, kSetExcept };
 
+enum LockMode { ForUpdate, ForNoKeyUpdate, ForShare, ForKeyShare };
+enum LockWaitPolicy { NoWait, SkipLocked, None };
```
Comment (mweisgut) [inline src/sql/SelectStatement.h:14]: How about `RowLockMode` and `RowLockWaitPolicy`?

File: src/parser/bison_parser.y [hunk 5]
```diff
@@ -1114,6 +1129,44 @@ alias : AS IDENTIFIER { $$ = new Alias($2); }
...
+| opt_locking_clause_list locking_clause {
+  $1->push_back($2);
+  $$ = $1;
+};
+
+locking_clause : FOR lock_strength opt_lock_waiting_policy {
+  $$ = new LockingClause();
+  $$->depTable = nullptr;
+  $$->lockMode = $2;
+  $$->lockWaitPolicy = $3;
```
Comment (mweisgut) [inline src/parser/bison_parser.y:1152]: `lockWaitPolicy` --> `rowLockWaitPolicy`?

File: src/sql/SelectStatement.h [hunk 6]
```diff
@@ -56,6 +59,12 @@ struct SetOperation {
   LimitDescription* resultLimit;
 };
 
+struct LockingClause {
+  LockMode lockMode;
+  LockWaitPolicy lockWaitPolicy;
+  std::vector<char*>* depTable;
```
Comment (mweisgut) [inline src/sql/SelectStatement.h:65]: `tables` or `target_tables` instead of `depTable`?

File: src/util/sqlhelper.cpp [hunk 7]
```diff
@@ -193,6 +193,34 @@ void printSelectStatementInfo(const SelectStatement* stmt, uintmax_t numIndent)
...
+        inprint("FOR UPDATE", numIndent + 3);
+      } else if (lockingClause->lockMode == LockMode::ForNoKeyUpdate) {
+        inprint("FOR NO KEY UPDATE", numIndent + 3);
+      } else if (lockingClause->lockMode == LockMode::ForShare) {
+        inprint("FOR SHARE", numIndent + 3);
+      } else if (lockingClause->lockMode == LockMode::ForKeyShare) {
+        inprint("FOR KEY SHARE", numIndent + 3);
+      }
+      if (lockingClause->depTable != nullptr) {
+        inprint("Target table:", numIndent + 2);
```
Comment (mweisgut) [inline src/util/sqlhelper.cpp:210]: `Target tables` (plural)?

File: test/queries/queries-bad.sql [hunk 8]
```diff
@@ -41,3 +41,7 @@
 # ON is not supported by postgres. We follow postgres here since the sql-92 standard does not specify index
 # implementation details.
 !DROP INDEX myindex ON mytable;
+!SELECT * FROM test WHERE val = 2 FOR KEY UPDATE;
+!SELECT * FROM test WHERE val = 2 FOR SHARE test1;
+!SELECT * FROM test WHERE val = 2 FOR NO KEY SHARE;
+!SELECT * FROM test WHERE val = 2 NOWAIT FOR UPDATE;
```
Comment (mweisgut) [inline test/queries/queries-bad.sql:47]: Missing newline at end of file

**Review by mweisgut — CHANGES_REQUESTED** [review]
Haven't reviewed the tests yet.

**Comment by chrisxu333** [comment]
@mweisgut Thank you for your review. I've pushed another commit and addressed them accordingly. Let me know if the testcases needs to change.

File: test/select_tests.cpp [hunk 9]
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
     }
   }
 }
+TEST(LockingClauseWithoutWaitPolicy) {
```
Comment (mweisgut) [inline test/select_tests.cpp:808]: Please add an empty line between the closing brace and the new test case.

Comment (mweisgut) [reply test/select_tests.cpp:808]: Please also check
- the number of elements in `lockings`
- `tables` (of the struct `LockingClause`)

File: test/select_tests.cpp [hunk 10]
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
...
+  stmt = (SelectStatement*)result.getStatement(2);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockMode, RowLockMode::ForNoKeyUpdate);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::None);
+
+  stmt = (SelectStatement*)result.getStatement(3);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockMode, RowLockMode::ForKeyShare);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::None);
+}
+
+TEST(LockingClauseWithWaitPolicy) {
```
Comment (mweisgut) [inline test/select_tests.cpp:834]: Please also check
- the number of elements in `lockings`
- `tables` (of the struct `LockingClause`)

File: test/select_tests.cpp [hunk 11]
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
...
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::SkipLocked);
+
+  stmt = (SelectStatement*)result.getStatement(6);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockMode, RowLockMode::ForNoKeyUpdate);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::SkipLocked);
+
+  stmt = (SelectStatement*)result.getStatement(7);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockMode, RowLockMode::ForKeyShare);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::SkipLocked);
+}
```
Comment (mweisgut) [inline test/select_tests.cpp:878]: Add an empty line between test cases.

File: test/select_tests.cpp [hunk 12]
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
...
+TEST(LockingClauseWithTableReference) {
```
Comment (mweisgut) [inline test/select_tests.cpp:879]: Please also check
- the number of elements in `lockings`
- the number of elements in `tables`

File: test/select_tests.cpp [hunk 13]
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
...
+  ASSERT_STREQ(stmt->lockings->at(0)->tables->at(1), "c");
+
+  stmt = (SelectStatement*)result.getStatement(2);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockMode, RowLockMode::ForNoKeyUpdate);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::NoWait);
+  ASSERT_STREQ(stmt->lockings->at(0)->tables->at(0), "t");
+  ASSERT_STREQ(stmt->lockings->at(0)->tables->at(1), "c");
+}
+
+TEST(MultipleLockingClause) {
```
Comment (mweisgut) [inline test/select_tests.cpp:905]: Please also check the number of elements in `lockings` and add a statement with multiple locking clauses where one clause has no tables specified, and two other clauses have tables specified but the tables and numbers of specified tables are different.

**Review by mweisgut — CHANGES_REQUESTED** [review]
A few minor change requests.

**Comment by chrisxu333** [comment]
I've added another commit to cover the change request about testcases. Plz take a look.

**Review by mweisgut — APPROVED** [review]

**Comment by mweisgut** [comment]
@chrisxu333 thank you again for your contribution.

**Pull request merged**